paimon-vfs/paimon-vfs-hadoop/src/main/java/org/apache/paimon/vfs/hadoop/PaimonVirtualFileSystem.java

```diff
@@ -0,0 +1,418 @@
+                                + " which is not in an existing table");
+            }
+            PositionOutputStream out =
+                    vfsTableObjectIdentifier
+                            .fileIO()
+                            .newOutputStream(vfsTableObjectIdentifier.getRealPath(), overwrite);
+            return new FSDataOutputStream(out, statistics);
+        }
+    }
+
+    @Override
+    public FSDataInputStream open(Path path, int bufferSize) throws IOException {
+        VFSIdentifier vfsIdentifier = vfsOperations.getVFSIdentifier(getVirtualPath(path));
+        if (vfsIdentifier instanceof VFSCatalogIdentifier) {
+            throw new IOException(
+                    "Cannot open file for virtual path " + path + " which is a catalog");
+        } else if (vfsIdentifier instanceof VFSDatabaseIdentifier) {
+            throw new IOException(
+                    "Cannot open file for virtual path " + path + " which is a database");
+        } else if (vfsIdentifier instanceof VFSTableRootIdentifier) {
+            throw new IOException(
+                    "Cannot open file for table level virtual path " + path + " which is a table");
+        } else {
+            VFSTableObjectIdentifier vfsTableObjectIdentifier =
+                    (VFSTableObjectIdentifier) vfsIdentifier;
+            if (!vfsTableObjectIdentifier.isTableExist()) {
+                throw new IOException(
+                        "Cannot open file for virtual path "
+                                + path
+                                + " which is not in an existing table");
+            }
+            VFSInputStream in =
+                    new VFSInputStream(
+                            vfsTableObjectIdentifier
+                                    .fileIO()
+                                    .newInputStream(vfsTableObjectIdentifier.getRealPath()));
+            return new FSDataInputStream(in);
+        }
+    }
+
+    @Override
+    public FSDataOutputStream append(Path f, int bufferSize, Progressable progress)
+            throws IOException {
+        throw new IOException("Append is not supported");
+    }
+
+    @Override
+    public boolean rename(Path src, Path dst) throws IOException {
+        VFSIdentifier srcVfsIdentifier = vfsOperations.getVFSIdentifier(getVirtualPath(src));
+        VFSIdentifier dstVfsIdentifier = vfsOperations.getVFSIdentifier(getVirtualPath(dst));
+        if (srcVfsIdentifier instanceof VFSCatalogIdentifier) {
+            throw new IOException("Cannot rename from virtual path " + src + " which is a catalog");
+        } else if (srcVfsIdentifier instanceof VFSDatabaseIdentifier) {
+            throw new IOException(
+                    "Cannot rename from virtual path " + src + " which is a database");
+        } else if (srcVfsIdentifier instanceof VFSTableRootIdentifier) {
+            throw new IOException("Cannot rename from virtual path " + src + " which is a table");
```

Comment:
We can support rename table.

Reply:
Same as delete api. Shall we support all types of table, or only support object table?

Reply:
I think it is ok to all tables.